### PR TITLE
[SPARK-9867][SQL] Move utilities for binary data into ByteArray

### PR DESCRIPTION
**Pull request by maropu**
The utilities such as Substring#substringBinarySQL and BinaryPrefixComparator#computePrefix for binary data are put together in ByteArray for easy-to-read.

### CODE REVIEW

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
@rxin Could you review this?
The unit tests failed though, ISTM these failures also happen in master.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/ByteArray.java [hunk 1]
```diff
@@ -29,4 +31,45 @@
...
+            << (56 - 8 * i);
+      }
+      return p;
+    }
+  }
+
+  public static byte[] subStringSQL(byte[] bytes, int pos, int len) {
+    // This pos calculation is according to UTF8String#subStringSQL
+    if (pos > bytes.length) {
+      return new byte[0];
```
Comment (rxin) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/ByteArray.java]: can we reuse an empty array?

Comment (maropu) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/ByteArray.java]: Good idea and fixed.

**Comment by maropu** [comment]
If no problem, could you merge this? cc: @rxin

**Comment by maropu** [comment]
cc: @rxin it's a just remainder.

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
Thanks. I've merged this.

**Pull request closed without merging**